### PR TITLE
Delay JS execution: delay script (2/3)

### LINKED ISSUE #2904: Delay JS execution: delay script (2/3)
Sub-task for epic #2899 

### New script to handle loading the delayed scripts
- Write the script to handle the loading of the delayed script after user interaction

### PR DESCRIPTION
**Pull request by hellofromtonya**
Closes #2904 

Adds the JavaScript file to:

1. Find all `<script>` elements with the `data-rocketlazyloadscript` attribute (the value holds the `src`).
2. Moves the attribute value to the `src` attribute.
3. Removes the data attribute.

The above is triggered when the user interacts with the page, i.e. on a window event.

## TODO:

- [x] Trigger on user interactivity.
- [x] Remove the `data-rocketlazyloadscript` attribute after triggering (not needed anymore).
- [x] Remove the listener from the `window` once triggered.
- [x] Test with base64.
- [x] Add tests. 👈  we don't have JS testing bootstrapped into the repo...yet. Using JSFiddle for now.
- [x] Review the listener options - what if passive is not supported in an older browser? [Added](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Safely_detecting_option_support)
- [x] Modify `gulpfile.js`: add a task for compiling single script.
To run:
```
gulp js:compile_single --script=lazyload-scripts.js
```
- [x] Optimized the new script: `assets/js/lazyload-scripts.min.js`.

## Tests:
Using JSFiddle as we don't yet have JUnit or another JS test suite in Rocket. [See it in action here](https://jsfiddle.net/hellofromtonya/79efrd2L/20/).

Browser Tests:
- [x] Chrome: Version 84.0.4147.89 (Official Build) (64-bit)
- [x] Safari: Version 13.1.2 (14609.3.5.1.5)
- [x] Firefox: Version 78.0.2 (64-bit)
- [x] Firefox Developer Edition: Version 80.0b4 (64-bit) 

## Notes:

- Use the optimized (transpiled and minified) `assets/js/lazyload-scripts.min.js` file.
- And then insert the minified code as in line `<script>`. Why? Faster. No external resource to request from the web server.

### CODE REVIEW

File: assets/js/layzload-scripts.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
+class RocketLazyLoadScripts {
+	options = { passive: true };
+	attrName = 'data-rocketlazyloadscript';
```
Comment (hellofromtonya) [inline assets/js/layzload-scripts.js:3]: This attribute name follows the suggestion made here https://github.com/wp-media/wp-rocket/issues/2899#issuecomment-669311394. It's easily changeable as a property in this class.

**Comment by Tabrisrp** [comment]
@hellofromtonya From the technical specs, the script should also handle inline script:

Initial script:

```
<script>(function(d, s, id) {
var js, fjs = d.getElementsByTagName(s)[0];
if (d.getElementById(id)) return;
js = d.createElement(s); js.id = id;
js.src = "https://connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.5";
fjs.parentNode.insertBefore(js, fjs);
}(document, 'script', 'facebook-jssdk'));</script>
```

Lazy version:

```
<script data-rocketlazyloadscript="data:text/javascript;base64,KGZ1bmN0aW9uKGQsIHMsIGlkKSB7DQogICAgdmFyIGpzLCBmanMgPSBkLmdldEVsZW1lbnRzQnlUYWdOYW1lKHMpWzBdOw0KICAgIGlmIChkLmdldEVsZW1lbnRCeUlkKGlkKSkgcmV0dXJuOw0KICAgIGpzID0gZC5jcmVhdGVFbGVtZW50KHMpOyBqcy5pZCA9IGlkOw0KICAgIGpzLnNyYyA9ICJodHRwczovL2Nvbm5lY3QuZmFjZWJvb2submV0L2VuX1VTL3Nkay5qcyN4ZmJtbD0xJnZlcnNpb249djIuNSI7DQogICAgZmpzLnBhcmVudE5vZGUuaW5zZXJ0QmVmb3JlKGpzLCBmanMpOw0KICB9KGRvY3VtZW50LCAnc2NyaXB0JywgJ2ZhY2Vib29rLWpzc2RrJykpOw=="></script>
```

And the lazyload script adds back the inline content inside the `<script>` tags. On the current version, I can see that it sets the `src` attribute, but not the content inside the script element.

**Comment by hellofromtonya** [comment]
@Tabrisrp In the PHP code, are we using [base64_encode()](https://www.php.net/manual/en/function.base64-encode.php) to convert?

**Comment by Tabrisrp** [comment]
This part of the code is not written yet, but that's the plan yes

**Comment by hellofromtonya** [comment]
With `<script src="">` the browser knows how to process base64 `src` as JavaScript code. 

For example: 

```html
<script data-rocketlazyloadscript="data:text/javascript;base64, Y29uc3QgdGV4dENvbnRlbnQgPSAnQmFzZTY0IGNvZGUgcmFuIGFzIEphdmFTY3JpcHQhIHcwMHQhJzsKCmNvbnN0IGVsZW0gPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCggJ3BhZ2UtdGl0bGUnICk7CmVsZW0uc2V0QXR0cmlidXRlKCAnc3R5bGUnLCAnYmFja2dyb3VuZC1jb2xvcjogcmVkOyBjb2xvcjogI2ZmZjsgcGFkZGluZzogNDBweDsgbWFyZ2luOiAyMHB4OycgKTsKZWxlbS50ZXh0Q29udGVudCA9IHRleHRDb250ZW50OwoKY29uc29sZS5sb2coIHRleHRDb250ZW50ICk7"></script>
```

Is base64 for this original JavaScript:

```js
const textContent = 'Base64 code ran as JavaScript! w00t!';

const elem = document.getElementById( 'page-title' );
elem.setAttribute( 'style', 'background-color: red; color: #fff; padding: 40px; margin: 20px;' );
elem.textContent = textContent;

console.log( textContent );
```

[You can see it run here](https://jsfiddle.net/hellofromtonya/79efrd2L/4/).

Converting it back to a string format could present encoding problems. There's risk of breaking websites.

I'm thinking we should validate the script with a series of different sizes scripts to validate there are no issues with staying in base64 format.

**Comment by hellofromtonya** [comment]
@Tabrisrp  asked the question:

What if the base64 encoded script has a variable that another script uses? Is that variable available to the other script?

It's a great question. Let's try it and see what happens.

[Here's the JSFiddle with it running](https://jsfiddle.net/hellofromtonya/79efrd2L/20/)

Let's walk through it:

The original inline JavaScript looked like this:
```html
<script>
const textContent = 'Base64 code ran as JavaScript! w00t!';

const elem = document.getElementById( 'page-title' );
elem.setAttribute( 'style', 'background-color: red; color: #fff; padding: 40px; margin: 20px;' );
elem.textContent = textContent;

let varAnotherScriptWillAccess = textContent;

console.log( textContent );
</script>
```
Notice there's a variable called `varAnotherScriptWillAccess`. When Rocket optimizes the above JS code by encoding it into base64, it's changed to this:

```html
<script data-rocketlazyloadscript="data:text/javascript;base64, Y29uc3QgdGV4dENvbnRlbnQgPSAnQmFzZTY0IGNvZGUgcmFuIGFzIEphdmFTY3JpcHQhIHcwMHQhJzsKCmNvbnN0IGVsZW0gPSBkb2N1bWVudC5nZXRFbGVtZW50QnlJZCggJ3BhZ2UtdGl0bGUnICk7CmVsZW0uc2V0QXR0cmlidXRlKCAnc3R5bGUnLCAnYmFja2dyb3VuZC1jb2xvcjogcmVkOyBjb2xvcjogI2ZmZjsgcGFkZGluZzogNDBweDsgbWFyZ2luOiAyMHB4OycgKTsKZWxlbS50ZXh0Q29udGVudCA9IHRleHRDb250ZW50OwoKbGV0IHZhckFub3RoZXJTY3JpcHRXaWxsQWNjZXNzID0gdGV4dENvbnRlbnQ7Cgpjb25zb2xlLmxvZyggdGV4dENvbnRlbnQgKTs="></script>
```

Imagine there's another script that uses this variable, for example:

```html
	<script>
		(() => {
			const btn = document.getElementById('base64-btn');
			const resultsElem = document.getElementById('base64-results');

			btn.addEventListener( 'click', () => {
				// Notice the variable here from the base64 encoded script.
				resultsElem.innerText = varAnotherScriptWillAccess;
				resultsElem.setAttribute( 'style', 'background-color: red; color: #fff; padding: 20px; margin: 20px;' );
			});
		})();
	</script>
```

[Run it the JSFiddle instance](https://jsfiddle.net/hellofromtonya/79efrd2L/20/). It works. 

Answer to Remy's question is:

Yes, other scripts can access variables from base64 scripts _when_ (this is important) they attempt to use it _after_ the script is loaded (after the user interacts with the window and our code shifts the data attribute back to `src`). If the other script tries to access it before that happens, it won't work. But that's not due to base64. That's because the JS is not in the DOM yet.

**Comment by hellofromtonya** [comment]
@Tabrisrp Commit https://github.com/wp-media/wp-rocket/pull/2969/commits/3c4608e2e3876e7cabf4eadb16d1b90c492a680b adds a check for if the browser supports passive options as well as the newer 3rd arg as an object or boolean. I used the code from this MDN web docs article:  ["Safely detecting option support"](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Safely_detecting_option_support).

Why did I move the properties into the constructor? Safari complained about it. This won't be a problem for customers as we'll need to run this script through a transpiler (Babel) to convert from ES6 to ES5.

👉  Note: Let's make sure that it does run through Babel when we go to build 3.7.

This PR is ready for review.

**Comment by hellofromtonya** [comment]
@Tabrisrp and @engahmeds3ed I added a new task to the `gulpfile.js` for running babel (transpiler) and uglify (minify) on a single script. Then optimized the new script.

The production script to use in Part 3 is `assets/js/lazyload-scripts.min.js`. 

FYI: To run the new gulp task, type:

```
gulp js:compile_single --script=lazyload-scripts.js
```

where the argv `--script=` tells gulp which file to optimize. It defaults to `./assets/js/` to save typing the whole thing.

**Pull request merged**